### PR TITLE
Implement support for custom blocks

### PR DESCRIPTION
**Pull request by JustTalDevelops**
Still needs code cleanup.

Below is an example of creating a custom skull block that can be rotated:

```go
type CustomSkull struct {
	empty
	transparent

	Facing cube.Direction
}

func (skull CustomSkull) Name() string {
	return "Custom Skull"
}

func (skull CustomSkull) Category() category.Category {
	return category.Nature()
}

func (skull CustomSkull) Textures() map[string]image.Image {
	return map[string]image.Image{
		"skull": skull.Texture(),
	}
}

func (skull CustomSkull) Texture() image.Image {
	texture, err := os.OpenFile("skull.png", os.O_RDONLY, os.ModePerm)
	if err != nil {
		panic(err)
	}
	defer texture.Close()
	img, err := png.Decode(texture)
	if err != nil {
		panic(err)
	}
	return img
}

func (skull CustomSkull) Geometry() []byte {
	data, err := os.ReadFile("skull.geo.json")
	if err != nil {
		panic(err)
	}
	return data
}

func (skull CustomSkull) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
	pos, _, used = firstReplaceable(w, pos, face, skull)
	if !used {
		return
	}

	skull.Facing = user.Rotation().Direction()
	place(w, pos, skull, user, ctx)
	return placed(ctx)
}

func (skull CustomSkull) EncodeItem() (name string, meta int16) {
	return "dragonfly:skull", 0
}

// EncodeBlock ...
func (skull CustomSkull) EncodeBlock() (string, map[string]any) {
	return "dragonfly:skull", map[string]any{"rotation": int32(skull.Facing)}
}

var skullHash = NextHash()

func (skull CustomSkull) Hash() uint64 {
	return skullHash | (uint64(skull.Facing) << 8)
}

func (skull CustomSkull) Properties() customblock.Properties {
	return customblock.Properties{
		CollisionBox: cube.Box(0.25, 0, 0.25, 0.75, 0.5, 0.75),
		SelectionBox: cube.Box(0.25, 0, 0.25, 0.75, 0.5, 0.75),
		Geometry:     "geometry.skull",
		Textures: map[string]customblock.Material{
			"*": customblock.NewMaterial("skull", customblock.OpaqueRenderMethod()),
		},
	}
}

func (skull CustomSkull) States() map[string][]any {
	return map[string][]any{
		"rotation": {int32(0), int32(1), int32(2), int32(3)},
	}
}

func (skull CustomSkull) Permutations() []customblock.Permutation {
	return []customblock.Permutation{
		{
			Condition: "query.block_state('rotation') == 1",
			Properties: customblock.Properties{
				Rotation: cube.Pos{0, 3, 0},
			},
		},
		{
			Condition: "query.block_state('rotation') == 2",
			Properties: customblock.Properties{
				Rotation: cube.Pos{0, 2, 0},
			},
		},
		{
			Condition: "query.block_state('rotation') == 3",
			Properties: customblock.Properties{
				Rotation: cube.Pos{0, 1, 0},
			},
		},
	}
}
```

```go
for _, direction := range cube.Directions() {
	world.RegisterBlock(block.CustomSkull{Facing: direction})
}
world.RegisterItem(block.CustomSkull{})
creative.RegisterItem(item.NewStack(block.CustomSkull{}, 1))

// Custom blocks need to be registered before the server is created

srv := conf.New()
```

### CODE REVIEW

File: server/block/customblock/material.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	texture string
+	// renderMethod is the method to use when rendering the material.
+	renderMethod RenderMethod
+	// faceDimming is if the material should be dimmed by the direction it's facing.
+	faceDimming bool
+	// ambientOcclusion is if the material should have ambient occlusion applied when lighting.
+	ambientOcclusion bool
+}
+
+// NewMaterial returns a new Material with the provided information. It also enabled face dimming and ambient occlusion
```
Comment (DaPigGuy) [inline server/block/customblock/material.go:15]: Second sentence seems to have some grammatical issues

File: server/block/customblock/model.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	"fmt"
+	"github.com/go-gl/mathgl/mgl32"
+	"github.com/go-gl/mathgl/mgl64"
+	"golang.org/x/exp/maps"
+)
+
+// Model represents the model of a custom block. It can contain multiple materials applied to different parts of the
+// model, as well as a reference to its geometry.
+type Model struct {
+	materials    map[string]Material
```
Comment (DaPigGuy) [inline server/block/customblock/model.go:13]: Should probably use MaterialTarget over a string?

File: server/block/customblock/render_method.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+// OpaqueRenderMethod returns the opaque rendering method for a material. It does not render an alpha layer, meaning it
+// does not support transparent or translucent textures, only textures that are fully opaque.
+func OpaqueRenderMethod() RenderMethod {
+	return RenderMethod{0}
+}
+
+// AlphaTestRenderMethod returns the alpha_test rendering method for a material. It does not allow for translucent
+// textures, only textures that are fully opaque or fully transparent, used for blocks such as regular glass. It also
+// disabled ambient occlusion by default.
```
Comment (DaPigGuy) [inline server/block/customblock/render_method.go:16]:
```suggestion
// disables ambient occlusion by default.
```

File: server/block/customblock/render_method.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	panic("unknown renderMethod type")
+}
+
+// AmbientOcclusion returns if ambient occlusion should be enabled by default for a material using this rendering method.
+func (m renderMethod) AmbientOcclusion() bool {
+	if m == 1 || m == 2 {
+		return false
+	}
+	return true
```
Comment (DaPigGuy) [inline server/block/customblock/render_method.go:60]:
```suggestion
	return m != 1 && m != 2
```

**Comment by JustTalDevelops** [comment]
This is now fully functional - feel free to review this. (although, I'm not exactly done with code cleanup yet, will do that ASAP though!)

**Comment by RoyalMCPE** [comment]
It maybe worth exporting [firstReplaceable](https://github.com/df-mc/dragonfly/blob/b191b9e4c23e961a9224e74492d229ad8f883a8b/server/block/block.go#L155) and [place](https://github.com/df-mc/dragonfly/blob/b191b9e4c23e961a9224e74492d229ad8f883a8b/server/block/block.go#L170) so that the end user doesn't have to copy the logic to add vanilla placement to their blocks.

**Comment by JustTalDevelops** [comment]
Yeah I do agree, I also think other fields such as `transparent` or `empty` could be exported, along with `newBreakInfo`, `newSmeltInfo`, etc.

File: server/world/block_state.go [hunk 5]
```diff
@@ -64,24 +68,49 @@ func init() {
...
 	rid := uint32(len(blocks))
+	blocks = append(blocks, unknownBlock{s})
+	if order {
+		sort.SliceStable(blocks, func(i, j int) bool {
+			nameOne, _ := blocks[i].EncodeBlock()
+			nameTwo, _ := blocks[j].EncodeBlock()
+			if nameOne == nameTwo {
+				return false
+			}
+			return fnv1.HashString64(nameOne) < fnv1.HashString64(nameTwo)
```
Comment (DaPigGuy) [inline server/world/block_state.go:86]:
```suggestion
			return nameOne == nameTwo && fnv1.HashString64(nameOne) < fnv1.HashString64(nameTwo)
```

File: server/server.go [hunk 6]
```diff
@@ -500,8 +509,16 @@ func (srv *Server) defaultGameData() minecraft.GameData {
 		GameRules:                    []protocol.GameRule{{Name: "naturalregeneration", Value: false}},
 		Difficulty:                   2,
 		Items:                        srv.itemEntries(),
+		CustomBlocks:                 srv.blockEntries(),
 		PlayerMovementSettings:       protocol.PlayerMovementSettings{MovementType: protocol.PlayerMovementModeServer, ServerAuthoritativeBlockBreaking: true},
 		ServerAuthoritativeInventory: true,
+		Experiments: []protocol.ExperimentData{
```
Comment (DaPigGuy) [inline server/server.go:515]: Is this still needed?

File: server/server.go [hunk 7]
```diff
@@ -675,6 +692,37 @@ func (srv *Server) itemEntries() (entries []protocol.ItemEntry) {
...
+}
+
+// blockEntries loads a list of all custom block entries of the server, ready to be sent in the StartGame packet. If the
+// list does not exist, it will be created and stored for future use.
+func (srv *Server) blockEntries() (entries []protocol.BlockEntry) {
+	if srv.blockComponents == nil {
+		srv.makeBlockComponents()
+	}
+	for name, properties := range srv.blockComponents {
+		pretty.Println(properties)
```
Comment (DaPigGuy) [inline server/server.go:717]: Remove this debug

File: server/internal/blockinternal/components.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		for condition, value := range levels {
+			builder.AddPermutation(condition, map[string]any{"minecraft:block_light_filter": map[string]any{
+				"lightLevel": int32(value),
+			}})
+		}
+	}
+	if i, ok := base.(block.Breakable); ok {
+		info := i.BreakInfo()
+		builder.AddComponent("minecraft:destroy_time", map[string]any{"value": float32(info.Hardness)})
+		// TODO: Explosion resistance.
```
Comment (DaPigGuy) [inline server/internal/blockinternal/components.go:80]: This can be added now

File: server/internal/blockinternal/components.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
+package blockinternal
+
+import (
+	"github.com/df-mc/dragonfly/server/block"
+	"github.com/df-mc/dragonfly/server/world"
+)
+
+// Components returns all the components of the given custom block group. If the group; has no components, a nil map
```
Comment (DaPigGuy) [inline server/internal/blockinternal/components.go:8]:
```suggestion
// Components returns all the components of the given custom block group. If the group has no components, a nil map
```

File: server/block/cube/axis.go [hunk 10]
```diff
@@ -35,7 +35,7 @@ func (a Axis) RotateLeft() Axis {
 
 // RotateRight rotates an Axis from X to Z or from Z to X.
 func (a Axis) RotateRight() Axis {
-	// No difference in rotating left or right for an Axis.
+	// No difference in rotating left or right for an Facing.
```
Comment (DaPigGuy) [inline server/block/cube/axis.go:38]: Is this change right? And should probably clarify that this is referring to the Y axis

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/block/cube/axis.go (hunk 10 above)
Comment (JustTalDevelops) [reply server/block/cube/axis.go:38]: No, probably a messy refactor.

**Pull request merged**